cclib/method/ddec.py

```diff
@@ -271,59 +281,102 @@ def calculate(self, indices=None, fupdate=0.05):
+        # continue to next step. Otherwise, calculate g and h. In both cases, calculate new
+        # rho_cond.
+        steps = 5
+        while steps < 7:
+            self.logger.info("Optimizing grid weights. (Step {}/7)".format(steps))
             self.N_A.append(self._calculate_w_and_u())
 
             # Update kappa as described in Figure S4.2 of doi: 10.1039/c6ra04656h
-            if numpy.any([x if x < -1e-5 else 0 for x in self.N_A[-1]]) and steps != 4:
+            if numpy.any([x if x < -1e-5 else 0 for x in self.N_A[-1]]):
                 update_kappa = True
             if (
-                steps > 5
-                and update_kappa
-                and numpy.diff(self.N_A)[-1] < 1e-5
-                and numpy.diff(self.N_A)[-2] < 1e-5
+                update_kappa
+                and numpy.any(numpy.diff(self.N_A)[-1] < 1e-5)  # change in N_A during last cycle
+                and numpy.any(numpy.diff(self.N_A)[-2] < 1e-5)  # change in N_A during last to
+                # to second cycle
             ):
                 update_kappa = False
-                kappa = [0.0 for x in self.data.atomnos]
+                self._kappa = [0.0 for x in self.data.atomnos]
 
-            # TODO: EXIT LOOP HERE for steps = 7
             if not update_kappa:
+                # Increment steps
                 steps = steps + 1
+                # Calculate G_A and H_A to update w_A based on S4.3 in doi: 10.1039/c6ra04656h
+                self.reshape_G()
+                self.calculate_H()
+
             else:
                 # `steps` not incremented in this case
                 # First update kappa based on equation 93
-                kappa_new = kappa - self.N_A / self.u_A
-                kappa = [x if x > 0 else 0.0 for x in kappa_new]
-                # TODO: In both cases, calculate G_A and H_A to update w_A based on S4.3 in doi: 10.1039/c6ra04656h
+                kappa_new = self._kappa - self.N_A[cycle] / self.u_A[cycle]
+                self._kappa = [x if x > 0 else 0.0 for x in kappa_new]
+
+            # Update weights (w_A) using equation 96 in doi: 10.1039/c6ra04656h
+            for atomi in range(len(self._cond_density)):
+                self._cond_density[atomi] = math.exp(self._kappa[atomi]) * self._h[atomi]
+
+            # Update rho_cond for next iteration
+            self._update_rho_cond()
+
+        # * STEP 7 *
+        # In step 7, calculate w and u. Then, stop to calculate refcharges.
+        self.logger.info("Optimizing grid weights. (Step 7/7)")
+        self.N_A.append(self._calculate_w_and_u())
+
+        # Finally, store calculated DDEC6 charges in fragcharges
+        self.logger.info("Creating fragcharges: array[1]")
+        self.fragcharges = numpy.array(self.data.atomnos - self.N_A[-1], dtype=float)
 
     def calculate_refcharges(self):
         """ Calculate reference charges from proatom density and molecular density
```

Comment:
Could you document in the docstring what this returns?

Reply:
I've added information about which quantities this function returns. Thanks!